dsn.go

```diff
@@ -13,73 +14,43 @@ type fireboltSettings struct {
 	accountName string
 }
 
-// splitString split string into two string, when first split char is encountered
-// return two strings:
-//    - first string before the encountered character, with removed backtick escape
-//    - second string is remaining string, including the character on which it was split
-func splitString(str string, splitChars []uint8) (string, string) {
-	var res string
-	for i := 0; i < len(str); i++ {
-		if str[i] == '\\' {
-			res += string(str[i+1])
-			i += 1
-			continue
-		}
-		for _, v := range splitChars {
-			if str[i] == v {
-				return res, str[i:]
-			}
-		}
-		res += string(str[i])
-	}
-	return res, ""
-}
-
-// parseRemainingDSN checks whether dsn has an expected prefix, removes it, and return two strings:
-//    - before the stopChars
-//    - after the stopChars, including the char itself
-func parseRemainingDSN(dsn string, expectedPrefix string, stopChars []uint8) (string, string, error) {
-	if !strings.HasPrefix(dsn, expectedPrefix) {
-		return "", dsn, fmt.Errorf("expected prefix not found: %s", expectedPrefix)
-	}
-
-	dsn = dsn[len(expectedPrefix):]
-	res, dsn := splitString(dsn, stopChars)
-	return res, dsn, nil
-}
+const dsnPattern = `^firebolt://(?P<username>.*@.*):(?P<password>.*)@(?P<database>\w+)(?:/(?P<engine>[^?]+))?(?:\?(?P<parameters>\w+\=[^=&]+(?:\&\w+=[^=&]+)*))?$`
+const paramsPattern = `(?P<key>\w+)=(?P<value>[^=&]+)`
 
 // ParseDSNString parses a dsn in a format: firebolt://username:password@db_name[/engine_name][?account_name=organization]
 // returns a settings object where all parsed values are populated
 // returns an error if required fields couldn't be parsed or if after parsing some characters were left unparsed
 func ParseDSNString(dsn string) (*fireboltSettings, error) {
 	var result fireboltSettings
+	dsnExpr := regexp.MustCompile(dsnPattern)
+	paramsExpr := regexp.MustCompile(paramsPattern)
 
-	var err error
-
-	// parse username
-	result.username, dsn, err = parseRemainingDSN(dsn, "firebolt://", []uint8{':'})
-	if err != nil {
-		return nil, ConstructNestedError("error during username parsing", err)
-	}
+	infolog.Println("Parsing DSN")
+	dsnMatch := dsnExpr.FindStringSubmatch(dsn)
 
-	// parse password
-	result.password, dsn, err = parseRemainingDSN(dsn, ":", []uint8{'@'})
-	if err != nil {
-		return nil, ConstructNestedError("error during password parsing", err)
+	if len(dsnMatch) == 0 {
+		return nil, errors.New("invalid connection string format")
 	}
 
-	// parse database
-	result.database, dsn, err = parseRemainingDSN(dsn, "@", []uint8{'/', '?'})
-	if err != nil {
-		return nil, ConstructNestedError("error during database parsing", err)
+	result.username = dsnMatch[1]
+	result.password = dsnMatch[2]
+	result.database = dsnMatch[3]
+	if len(dsnMatch[4]) > 0 {
+		// engine name was provided
+		result.engineName = dsnMatch[4]
 	}
```

Comment:
Is there a way to leverage named capture groups in the regex for access here? e.g. `dsnMatch.username` would be safer than `dsnMatch[1]`

Reply:
matcher provider names as a separate list. We would still need to access the value by index